File: current_session/529.py
"""
529
"""
from typing import List
class Solution:
    def count(self, board, i, j):
        nbs = [(1,0),(-1,0),(0,1),(0,-1),(1,1),(-1,-1),(-1,1),(1,-1)]
        cnt = 0
        for a,b in nbs:
            if 0<=i+a<len(board) and 0<=j+b<len(board[0]) and board[i+a][j+b] == 'M':
                cnt += 1
        return cnt

    def updateBoard(self, board: List[List[str]], click: List[int]) -> List[List[str]]:
        # 1st pass: see if it's gonna bomb or not
        # 
        # 2nd pass: generate number and blank
        #           see if it's on a number or not
        #           if it's on a blank, reveal blank first and reveal number as well

        i, j = click[0], click[1]
        if board[i][j] == 'M':
            board[i][j] = 'X'
            return board
        
        cntAtClick = self.count(board, i, j)
        if cntAtClick == 0:
            # reveal all connected '0's and numbers as well
            st = [(i,j)]
            toChange = []
            # print(board)
            visited = set()
            while st:
                p, q = st.pop()
                visited.add((p,q))
                # print('p,q:',p,q)
                cnt = self.count(board, p, q)
                if cnt == 0:
                    toChange += (p, q, 'B'),
                    for a, b in [(p+1,q),(p-1,q),(p,q+1),(p,q-1),(p+1,q+1),(p+1,q-1),(p-1,q+1),(p-1,q-1)]:
                        if 0<=a<len(board) and 0<=b<len(board[0]) and (a,b) not in visited:
                            st.append((a,b))
                else:
                    toChange += (p, q, str(cnt)),
            
            while toChange:
                r, c, val = toChange.pop()
                board[r][c] = val
            return board
        
        else:
            board[i][j] = str(cntAtClick)
            return board

File: current_session/c++/529.cpp
#include <iostream>
#include <vector>
#include <set>

using namespace std;

class Solution {
 public:
  vector<vector<char>> updateBoard(vector<vector<char>>& board, vector<int>& click) {
    // 1st pass: change the click into 'X' if it's mine
    // 2nd pass: dfs, count its nbs
    // and stop searching when we hit "digits"

    if (board[click[0]][click[1]] == 'M') {
      // modify the place to 'X', return
      board[click[0]][click[1]] = 'X';
      return board;
    }

    dfs(click[0], click[1], board);
    return board;
  }
 private:

  bool valid(int i, int j, const vector<vector<char>>& board) {
    return (0<=i && i<board.size() && 0<=j && j<board[0].size());
  }

  // click the point
  // if blank: check its 8 nbs
  // if number: simply change the board to number
  void dfs(int i, int j, vector<vector<char>>& board) {
    // only handle inbound unrevealed squares
    if (valid(i,j,board) && board[i][j] == 'E') {
      // count the nbs
      int cnt = 0;
      if (valid(i-1,j+1,board) && board[i-1][j+1] == 'M') cnt += 1;
      if (valid(i-1,j-1,board) && board[i-1][j-1] == 'M') cnt += 1;
      if (valid(i+1,j+1,board) && board[i+1][j+1] == 'M') cnt += 1;
      if (valid(i+1,j-1,board) && board[i+1][j-1] == 'M') cnt += 1;
      if (valid(i-1,j,board) && board[i-1][j] == 'M') cnt += 1;
      if (valid(i+1,j,board) && board[i+1][j] == 'M') cnt += 1;
      if (valid(i,j-1,board) && board[i][j-1] == 'M') cnt += 1;
      if (valid(i,j+1,board) && board[i][j+1] == 'M') cnt += 1;
      
      // either blank or digit
      if (cnt == 0) {
        board[i][j] = 'B';
        dfs(i-1,j+1,board);
        dfs(i-1,j-1,board);
        dfs(i+1,j+1,board);
        dfs(i+1,j-1,board);
        dfs(i-1,j,board);
        dfs(i+1,j,board);
        dfs(i,j-1,board);
        dfs(i,j+1,board);
      } else {
        board[i][j] = '0'+cnt;
      } 
    }
    return;
  }
};